README.md
# react-simple-tooltip

[![npm package][npm-badge]][npm]
[![Travis][build-badge]][build]
[![Codecov][codecov-badge]][codecov]
![Module formats][module-formats]

A lightweight and simple tooltip component for React

## Getting started

[![react-simple-tooltip](https://nodei.co/npm/react-simple-tooltip.png?downloads=true&downloadRank=true&stars=true)](https://nodei.co/npm/react-simple-tooltip/)

You can download `react-simple-tooltip` from the NPM registry via the `npm` or `yarn` commands

```shell
yarn add react-simple-tooltip
npm install react-simple-tooltip --save
```

If you don't use package manager and you want to include `react-simple-tooltip` directly in your html, you could get it from the UNPKG CDN

```html
https://unpkg.com/react-simple-tooltip/dist/react-simple-tooltip.min.js.
```

## Usage

### Attached to a Component

```javascript
import React from "react"
import Tooltip from "react-simple-tooltip"

const App = () => (
  <Tooltip content="😎">
    <button>Hover me !</button>
  </Tooltip>
)
```

### Standalone

```javascript
import React from "react"
import Tooltip from "react-simple-tooltip"

const App = () => (
  <div style={{position: "relative"}}>
    <Tooltip
      style={{position: "absolute", top: "50%", right: "0"}}
      content="😎"
    />
  </div>
)
```

### Custom css

```javascript
import React from "react"
import Tooltip from "react-simple-tooltip"
import {css} from "styled-components"

const App = () => (
  <Tooltip
    content="😎"
    customCss={css`
      white-space: nowrap;
    `}
  >
    <button>Hover me !</button>
  </Tooltip>
)
```

## Demo

See [Demo page][github-page]

## Props

| Name         | PropType                                            | Description                        | Default   |
| ------------ | --------------------------------------------------- | ---------------------------------- | --------- |
| arrow        | PropTypes.number                                    | Arrow size, in pixels              | 8         |
| background   | PropTypes.string                                    | Tooltip background color           | "#000"    |
| border       | PropTypes.string                                    | Tooltip border color               | "#000"    |
| color        | PropTypes.string                                    | Tooltip text color                 | "#fff"    |
| content      | PropTypes.any.isRequired                            | Tooltip content                    | -         |
| customCss    | PropTypes.any                                       | Custom css                         | -         |
| fadeDuration | PropTypes.number                                    | Fade duration, in milliseconds     | 0         |
| fadeEasing   | PropTypes.string                                    | Fade easing                        | "linear"  |
| fixed        | PropTypes.bool                                      | Tooltip behavior, hover by default | false     |
| fontFamily   | PropTypes.bool                                      | Tooltip text font-family           | "inherit" |
| fontSize     | PropTypes.bool                                      | Tooltip text font-size             | "inherit" |
| padding      | PropTypes.number                                    | Tooltip padding, in pixels         | 16        |
| placement    | PropTypes.oneOf(["left", "top", "right", "bottom"]) | Tooltip placement                  | "top"     |
| radius       | PropTypes.number                                    | Tooltip border radius              | 0         |
| zIndex       | PropTypes.number                                    | Tooltip z-index                    | 1         |

## Contributing

- ⇄ Pull/Merge requests and ★ Stars are always welcome.
- For bugs and feature requests, please [create an issue][github-issue].
- Pull requests must be accompanied by passing automated tests (`npm test`).

See [CONTRIBUTING.md](./CONTRIBUTING.md) guidelines

## Changelog

See [CHANGELOG.md](./CHANGELOG.md)

## License

This project is licensed under the MIT License - see the [LICENCE.md](./LICENCE.md) file for details

[npm-badge]: https://img.shields.io/npm/v/react-simple-tooltip.svg?style=flat-square
[npm]: https://www.npmjs.org/package/react-simple-tooltip
[build-badge]: https://img.shields.io/travis/xuopled/react-simple-tooltip/master.svg?style=flat-square
[build]: https://travis-ci.org/xuopled/react-simple-tooltip
[codecov-badge]: https://img.shields.io/codecov/c/github/xuopled/react-simple-tooltip.svg?style=flat-square
[codecov]: https://codecov.io/gh/xuopled/react-simple-tooltip
[module-formats]: https://img.shields.io/badge/module%20formats-umd%2C%20cjs%2C%20esm-green.svg?style=flat-square
[github-page]: https://xuopled.github.io/react-simple-tooltip
[github-issue]: https://github.com/xuopled/react-simple-tooltip/issues/new


CHANGELOG.md
# 2.6.1 — 2019-07-25

- Fixed: ios touched devices by @vnoitkumar

# 2.6.0 — 2019-05-23

- Updated: Replacing styled-components with emotion/core

# 2.5.0 — 2019-03-08

- Added: `customCss` prop

# 2.4.0 — 2019-03-08

- Updated: dependencies
- Updated: tests

# 2.3.3 — 2018-11-08

- Fixed: z-index prop
- Updated: Dependencies

# 2.3.2 — 2018-06-13

- Removes: index.css in package.json

# 2.3.1 — 2018-03-12

- Fixed: Removed warning when onMouseEnter and Leave are booleans instead of undefined by @vincentdesmares

# 2.3.0 - 2018-01-04

- Updated: FadeDuration to be in ms,
- Added: Offset prop to allow spacing between arrow and trigger

# 2.2.0 - 2018-01-03

- Added: Fade animation props (by @BenLorantfy)

# 2.1.0 - 2017-12-08

- Added: Standalone version

# 2.0.0 - 2017-10-18

- Added: Nwb build
- Updated: Nwb build
- Added: Tests
- Added: styled-components dependency
- Removed: classnames dependency

# 1.0.6 - 2017-08-24

- Updated: Import PropTypes from `prop-types` package
- Added: Yarn

# 1.0.5 - 2015-07-09

- Updated: webpack build

# 1.0.4 - 2015-07-08

- Added: webpack build

# 1.0.3 - 2015-07-03

- Added: fixed tooltip example

# 1.0.2 - 2015-07-02

- Updated: tooltips example screenshot

# 1.0.1 - 2015-07-02

- Updated: repository url in package.json
- Added: Npm Badge in README

# 1.0.0 - 2015-07-01

- Initial release


src/index.js
/** @jsx jsx */
import React from "react"
import PropTypes from "prop-types"
import {css, jsx} from "@emotion/core"

import Arrow from "./components/Arrow"
import Tooltip from "./components/Tooltip"
import Bubble from "./components/Bubble"
import {easingPropType} from "./utils/propTypes"

const ContainerCss = css`
  position: relative;
  display: inline-block;
`

class Wrapper extends React.Component {
  constructor() {
    super()

    this.state = {
      open: false,
    }

    this.handleMouseEnter = this.handleMouseEnter.bind(this)
    this.handleMouseLeave = this.handleMouseLeave.bind(this)
    this.handleTouch = this.handleTouch.bind(this)
  }

  handleMouseEnter() {
    this.setState({open: true})
  }

  handleMouseLeave() {
    this.setState({open: false})
  }

  handleTouch() {
    const isOpen = this.state.open
    this.setState({open: !isOpen})
  }

  render() {
    const {open} = this.state
    const {
      arrow,
      background,
      border,
      children,
      color,
      content,
      customCss,
      fadeDuration,
      fadeEasing,
      fixed,
      fontFamily,
      fontSize,
      offset,
      padding,
      placement,
      radius,
      zIndex,
      ...props
    } = this.props
    const hasTrigger = children !== undefined && children !== null
    const tooltipElement = (
      <Tooltip
        open={!hasTrigger || fixed ? true : open}
        placement={placement}
        offset={offset + arrow}
        zIndex={zIndex}
        fadeEasing={fadeEasing}
        fadeDuration={fadeDuration}
      >
        <Bubble
          background={background}
          border={border}
          color={color}
          radius={radius}
          fontFamily={fontFamily}
          fontSize={fontSize}
          padding={padding}
        >
          <Arrow
            width={arrow}
            background={background}
            border={border}
            color={color}
            placement={placement}
          />
          {content}
        </Bubble>
      </Tooltip>
    )
    return hasTrigger ? (
      <div
        onMouseEnter={!fixed ? this.handleMouseEnter : undefined}
        onMouseLeave={!fixed ? this.handleMouseLeave : undefined}
        onTouchEnd={!fixed ? this.handleTouch : undefined}
        css={css`
          ${ContainerCss};
          ${customCss}
        `}
        {...props}
      >
        {children}
        {tooltipElement}
      </div>
    ) : (
      <div
        css={css`
          ${ContainerCss};
          ${customCss}
        `}
        {...props}
      >
        {tooltipElement}
      </div>
    )
  }
}

Wrapper.propTypes = {
  arrow: PropTypes.number,
  background: PropTypes.string,
  border: PropTypes.string,
  children: PropTypes.any,
  color: PropTypes.string,
  content: PropTypes.any.isRequired,
  customCss: PropTypes.any,
  fadeDuration: PropTypes.number,
  fadeEasing: easingPropType,
  fixed: PropTypes.bool,
  fontFamily: PropTypes.string,
  fontSize: PropTypes.string,
  offset: PropTypes.number,
  padding: PropTypes.number,
  placement: PropTypes.oneOf(["left", "top", "right", "bottom"]),
  radius: PropTypes.number,
  zIndex: PropTypes.number,
}

Wrapper.defaultProps = {
  arrow: 8,
  background: "#000",
  border: "#000",
  children: null,
  color: "#fff",
  fadeDuration: 0,
  fadeEasing: "linear",
  fixed: false,
  fontFamily: "inherit",
  fontSize: "inherit",
  offset: 0,
  padding: 16,
  placement: "top",
  radius: 0,
  zIndex: 1,
}

Wrapper.displayName = "Tooltip.Wrapper"
Tooltip.displayName = "Tooltip"
Bubble.displayName = "Tooltip.Bubble"
Arrow.displayName = "Tooltip.Arrow"

export default Wrapper
